src/main/java/com/rackspacecloud/metrics/ingestionservice/influxdb/GCLineProtocolBackupService.java

```diff
@@ -0,0 +1,133 @@
+
+@Service
+@EnableConfigurationProperties(BackupProperties.class)
+public class GCLineProtocolBackupService implements LineProtocolBackupService {
+
+    // Reusable thread-safe date-time formatter for files going in a bucket
+    private static final DateTimeFormatter dateBucketFormat = DateTimeFormatter.ofPattern("yyyyMMdd")
+            .withZone(ZoneOffset.UTC);
+    private static final Logger LOGGER = LoggerFactory.getLogger(GCLineProtocolBackupService.class);
+    // The underlying blob storage. Dev/test will uses in-memory storage
+    private final Storage storage;
+    // The bucket where blobs will go
+    private final String cloudOutputBucket;
+
+    // Internal methods of this class attempt to talk to the Cache proxy
+    // instead of the GCLineProtocolBackupService object
+    @Resource
+    private LineProtocolBackupService self;
+
+    // To autowire Storage in prod we'll need to get an account and point
+    // GOOGLE_APPLICATION_CREDENTIALS to the json creds file
+    // for a service account
+    // as per https://www.baeldung.com/java-google-cloud-storage
+    @Autowired
+    public GCLineProtocolBackupService(Storage storage, BackupProperties backupProperties) {
+        Assert.notNull(storage, "Storage must not be null");
+        Assert.notNull(backupProperties, "The backup properties must not be null");
+        // TODO: remove gcs prefix to make abstraction cleaner
+        Assert.notNull(backupProperties.getGcsBackupBucket(), "The output bucket must not be null");
+        this.storage = storage;
+        this.cloudOutputBucket = backupProperties.getGcsBackupBucket();
+    }
+
+    /**
+     * We want data to be stored in the following format: /[profile-specific
+     * bucket]/[instance name]/[database name]/[retention policy
+     * name]/[YYYYMMDD]/[UUID of file].gz
+     *
+     * File is line-protocol, gzipped, size may require some testing
+     *
+     * @param location the bucket location (directory) of the file we are writing to
+     * @return the cached reference to the gzip output stream for that blob
+     */
+    @Cacheable(cacheNames = "lineProtocolBackupWriter")
+    public GZIPOutputStream getBackupStream(String location) throws IOException {
+        Assert.notNull(location, "location must not be null");
+        Assert.notNull(cloudOutputBucket, "bucket name was not injected properly");
+        Assert.notNull(storage, "storage was not injected properly");
+        String fileName = location + "/" + UUID.randomUUID().toString() + ".gz";
+        BlobId blobId = BlobId.of(cloudOutputBucket, fileName);
+        BlobInfo blobInfo = BlobInfo.newBuilder(blobId).setContentType("application/gzip").build();
+        // enable flushing the compressor
+        return new GZIPOutputStream(Channels.newOutputStream(storage.writer(blobInfo)), true);
+    }
+
+    @Override
+    public void writeToBackup(String payload, String instance, String database, String retentionPolicy)
+            throws IOException {
```

Comment:
Given this is a `public` method, please add input validation for the parameters. I would recommend using `IllegalArgumentException` here. I see that if `payload` is `null`, you will get `NullPointerException` on `line# 91`